### PR TITLE
Refactor insert / remove / replace / merge behavior

### PR DESCRIPTION
**Pull request by inyono**
Storybook: https://storybook.edtr.io/refactor-insert-remove/

We're getting there, still has some rough edges, though. Please try this out and report any (especially text) behavior that is different from current stable release. Also, feel free to review & feedback the change to the public API (so everything in /api)

## Known issues

- [x] ~store: after removing an child, check if its parent is empty. If so, also remove the parent~ #286
- [x] plugin-text: splitting blocks doesn't work correctly anymore
- [x] ~plugin-text: merging blocks just merges the documents, not the blocks~ #285
- [x] plugin-text: suggestions don't work
- [x] plugin-image: `isEmpty` check seems broken

### CODE REVIEW

File: api/store.api.md [hunk 1]
```diff
@@ -332,6 +389,12 @@ export const isEmpty: Selector<boolean, [string]>;
 // @public
 export const isFocused: Selector<boolean, [string]>;
 
+// @public (undocumented)
+export const mayInsertChild: import("..").Selector<boolean, [string]>;
```
Comment (inyono) [inline api/store.api.md]: That looks like it might break the build, add explicit type

File: api/store.api.md [hunk 2]
```diff
@@ -332,6 +389,12 @@ export const isEmpty: Selector<boolean, [string]>;
...
+
+// @public (undocumented)
+export const mayRemoveChild: import("..").Selector<boolean, [string]>;
```
Comment (inyono) [inline api/store.api.md]: That looks like it might break the build, add explicit type

File: packages/plugins/rows/src/index.tsx [hunk 3]
```diff
@@ -87,6 +87,28 @@ export function createRowsPlugin({
...
+        if (index === -1) return null
+        return index + 1
+      }
+    },
+
+    removeChild(state, id) {
+      const index = R.findIndex(child => child.id === id, state)
+      if (index === -1) return
+      state.remove(index)
+    }
```
Comment (inyono) [inline packages/plugins/rows/src/index.tsx]: We could extract that into its own state type `children`

File: api/core.api.md [hunk 4]
```diff
@@ -43,8 +43,6 @@ export interface DocumentProps {
     // (undocumented)
     pluginProps?: {
         config?: {};
-        insert?: DocumentEditorProps['insert'];
-        remove?: DocumentEditorProps['remove'];
```
Comment (inyono) [inline api/core.api.md]: Breaking change

File: api/internal__document-editor.api.md [hunk 5]
```diff
@@ -13,13 +13,8 @@ export interface DocumentEditorProps {
     focused: boolean;
     hasSettings: boolean;
     hasToolbar: boolean;
-    insert?(options?: {
-        plugin: string;
-        state?: unknown;
-    }): void;
```
Comment (inyono) [inline api/internal__document-editor.api.md]: Breaking change

File: api/internal__document-editor.api.md [hunk 6]
```diff
@@ -13,13 +13,8 @@ export interface DocumentEditorProps {
...
     // (undocumented)
     PluginToolbar: PluginToolbar;
-    remove?(): void;
```
Comment (inyono) [inline api/internal__document-editor.api.md]: Breaking change

File: api/internal__plugin-state.api.md [hunk 7]
```diff
@@ -16,14 +16,10 @@ export interface PluginProps {
     // (undocumented)
     config?: {};
     // (undocumented)
-    insert?: DocumentEditorProps['insert'];
-    // (undocumented)
```
Comment (inyono) [inline api/internal__plugin-state.api.md]: Breaking change

File: api/internal__plugin-state.api.md [hunk 8]
```diff
@@ -16,14 +16,10 @@ export interface PluginProps {
...
     name?: string;
     // (undocumented)
     parent?: PluginProps;
     // (undocumented)
-    remove?: DocumentEditorProps['remove'];
-    // (undocumented)
```
Comment (inyono) [inline api/internal__plugin-state.api.md]: Breaking change

File: api/internal__plugin.api.md [hunk 9]
```diff
@@ -8,18 +8,25 @@ import * as React from 'react';
...
     Component: React.ComponentType<EditorPluginProps<S, Config>>;
     config: Config | ((theme: Theme) => Config);
-    isEmpty?(state: StateTypeValueType<S>): boolean;
+    insertChild?(state: StateTypeReturnType<S>, { previousSibling, document }: {
+        previousSibling?: string;
+        document?: {
+            plugin: string;
+            state?: unknown;
+        };
+    }): void;
```
Comment (inyono) [inline api/internal__plugin.api.md]: Breaking change, replaces the previous `insert` in the plugin props

File: api/internal__plugin.api.md [hunk 10]
```diff
@@ -8,18 +8,25 @@ import * as React from 'react';
...
+    isEmpty?(state: StateTypeReturnType<S>): boolean;
```
Comment (inyono) [inline api/internal__plugin.api.md]: Breaking change, received the value type previously

File: api/internal__plugin.api.md [hunk 11]
```diff
@@ -8,18 +8,25 @@ import * as React from 'react';
...
     onKeyDown?(e: KeyboardEvent): boolean;
     onPaste?(data: DataTransfer): void | {
         state?: StateTypeSerializedType<S>;
     };
+    removeChild?(state: StateTypeReturnType<S>, id: string): void;
```
Comment (inyono) [inline api/internal__plugin.api.md]: Breaking change, replaces the previous `remove` in the plugin props

File: api/internal__plugin.api.md [hunk 12]
```diff
@@ -30,8 +37,6 @@ export interface EditorPluginProps<S extends StateType = StateType, Config exten
     editable: boolean;
     focused: boolean;
     id: string;
-    // @deprecated
-    name: string;
```
Comment (inyono) [inline api/internal__plugin.api.md]: Breaking change

File: api/plugin-text.api.md [hunk 13]
```diff
@@ -34,10 +32,11 @@ export interface BlockRendererProps {
 }
 
 // @public (undocumented)
-export function createTextPlugin({ placeholder, registry, theme }: {
+export function createTextPlugin({ placeholder, registry, theme, blockquote }: {
```
Comment (inyono) [inline api/plugin-text.api.md]: Slightly breaking change. Text Plugin now accepts the type of blockquote plugin as part of its config. If it's not provided, we don't show the blockquote controls.

File: api/internal__plugin-state.api.md [hunk 14]
```diff
@@ -33,7 +29,7 @@ export interface StateType<S = any, T = S, R = unknown> {
     createInitialState(helpers: StoreDeserializeHelpers): T;
     deserialize(serialized: S, helpers: StoreDeserializeHelpers): T;
     getFocusableChildren(state: T): FocusableChild[];
-    init(state: T, onChange: (initial: StateUpdater<T>, executor?: StateExecutor<StateUpdater<T>>) => void, pluginProps?: PluginProps): R;
+    init(state: T, onChange: (initial: StateUpdater<T>, executor?: StateExecutor<StateUpdater<T>>) => void): R;
```
Comment (inyono) [inline api/internal__plugin-state.api.md]: Breaking Change

**Comment by Knorrke** [comment]
- Plugin Inserts with `/...` don't work

File: api/plugin-text.api.md (hunk 13 above)
Comment (Knorrke) [reply api/plugin-text.api.md]: nice

**Comment by Knorrke** [comment]
Focus previous / next when empty plugin gets removed (by backspace and del respectively)

**Comment by Knorrke** [comment]
- Up and Down Navigation is weird (but also not working that well in the current demo)
cursor-navigation

**Pull request merged**